src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -442,6 +443,36 @@ public override BoundNode VisitThisReference(BoundThisReference node)
                     return base.VisitThisReference(node);
                 }
 
+                public override BoundNode VisitGotoStatement(BoundGotoStatement node)
+                {
+                    // Any jumping construct (for, break, if etc.) generates a BoundGotoStatement, so check this is actually a real goto statement
```

Comment:
Wait -- this is perfect! For some reason I thought local rewriting happened after closure conversions, so I didn't think we would have the generated label statements. Since all we need to do is find if the environment is in-between a backwards branch, why not keep a list of backward branches and then check to see if the closure location intersects any of those. If it does, don't do the optimization.

This would avoid all of my worry about new loop types breaking the optimization, since they'll necessarily have to introduce a new generated label and branch for the rest of the compiler to work.

Reply:
It has to be that the jump goes from an area after the closure is declared to a point between where the parent closure is declared and the closure is declared, or from a point in between the two closures to a point before the parent closure for there to be an issue.

A jump which goes from after the child closure to before the parent closure is fine, and a jump from before the child closure to after the parent closure is also fine

Reply:
Yeah, I think that's the precise statement. You could implement that, or you could try something simpler, like just checking if there's a label in-between the parent and the child environments. That would be a conservative approximation, right?

Reply:
I will look into it, and see what is doable

Reply:
The tests have now been updated.

At the moment the algorithm to decide whether it is semantically safe to merge closure environments is as perfect as it could be without doing control flow analysis.

With control flow  analysis, more could be done.

For example

```csharp
using System;

public static class Program
{
	public static void Main()
	{
		int a = 0;
        target: ;
		{
			int b = 1;
			Action action = () => Console.WriteLine(a + b);
			action();
		}

        return;

        goto target; // This can never be reached, so it is actually safe to merge closure environments
	}
}
```

Similarly the algorithm to decide whether merging closure environments will effect GC is as good as it could be without doing Escape analysis.

With escape analysis, more could be done as well.

For example

```csharp
using System;

public static class Program
{
	public static void Main()
	{
        int a = 0;
        {
            string s = "Hello";
            Action b = () => Console.WriteLine(s + a);
        }
        {
            string s = "World";
            Action b = () => Console.WriteLine(s + a);
        }
	}
}
```

Currently the three display classes will not be merged into one. However since they all have the same lifetime, there is no reason why they should not be merged in theory.